File: scripts/check-portable-api-client.py
#!/usr/bin/env python3
"""Fast structural validation for Ugoite's portable API client boundary."""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


def fail(message: str) -> None:
    raise AssertionError(message)


def quoted_items(block: str) -> list[str]:
    return re.findall(r'"([a-z][a-z0-9_.]+)"', block)


def rust_manifest_operations(source: str) -> list[str]:
    match = re.search(
        r"pub const SUPPORTED_OPERATIONS: &\[&str\] = &\[(.*?)\];",
        source,
        re.S,
    )
    if not match:
        fail("Rust SUPPORTED_OPERATIONS manifest was not found")
    return quoted_items(match.group(1))


def typescript_manifest_operations(source: str) -> list[str]:
    match = re.search(
        r"export const UGOITE_API_OPERATIONS = \[(.*?)\] as const;",
        source,
        re.S,
    )
    if not match:
        fail("TypeScript UGOITE_API_OPERATIONS manifest was not found")
    return quoted_items(match.group(1))


def frontend_operation_uses() -> dict[str, list[str]]:
    pattern = re.compile(
        r"protocolFetch(?:<.*?>)?\s*\(\s*[\"']([^\"']+)",
        re.S,
    )
    uses: dict[str, list[str]] = {}
    for path in sorted((ROOT / "frontend/src/lib").glob("*-api.ts")):
        for operation in pattern.findall(path.read_text()):
            uses.setdefault(operation, []).append(str(path.relative_to(ROOT)))
    return uses


def cli_operation_uses() -> dict[str, list[str]]:
    pattern = re.compile(
        r"http::execute\s*\(\s*[^,]+,\s*[\"']([^\"']+)",
        re.S,
    )
    uses: dict[str, list[str]] = {}
    for path in sorted((ROOT / "crates/ugoite-cli/src/commands").glob("*.rs")):
        for operation in pattern.findall(path.read_text()):
            uses.setdefault(operation, []).append(str(path.relative_to(ROOT)))
    return uses


def dependency_keys_from_manifest(path: Path) -> set[str]:
    dependencies: set[str] = set()
    in_dependencies = False
    for raw_line in path.read_text().splitlines():
        line = raw_line.strip()
        if line.startswith("["):
            in_dependencies = line == "[dependencies]"
            continue
        if not in_dependencies or not line or line.startswith("#") or "=" not in line:
            continue
        key = line.split("=", 1)[0].strip()
        if key:
            dependencies.add(key)
    return dependencies


def check_cargo_manifests() -> None:
    for path in ROOT.glob("**/Cargo.toml"):
        path.read_text()
    (ROOT / "Cargo.lock").read_text()

    api_dependencies = dependency_keys_from_manifest(
        ROOT / "crates/ugoite-api-client/Cargo.toml"
    )
    forbidden_api = {
        "axum",
        "opendal",
        "reqwest",
        "tokio",
        "ugoite-core",
        "ugoite-domain",
        "ugoite-storage",
        "wasm-bindgen",
        "web-sys",
    }
    unexpected = sorted(api_dependencies & forbidden_api)
    if unexpected:
        fail(f"ugoite-api-client has forbidden dependencies: {unexpected}")

    wasm_dependencies = dependency_keys_from_manifest(
        ROOT / "crates/ugoite-wasm/Cargo.toml"
    )
    forbidden_wasm = {
        "opendal",
        "reqwest",
        "tokio",
        "ugoite-core",
        "ugoite-storage",
    }
    unexpected = sorted(wasm_dependencies & forbidden_wasm)
    if unexpected:
        fail(f"ugoite-wasm has forbidden dependencies: {unexpected}")


def check_source_boundaries() -> None:
    for path in sorted((ROOT / "frontend/src/lib").glob("*-api.ts")):
        content = path.read_text()
        if "apiFetch" in content:
            fail(f"{path.relative_to(ROOT)} imports or calls apiFetch directly")

    for path in sorted((ROOT / "crates/ugoite-cli/src/commands").glob("*.rs")):
        content = path.read_text()
        forbidden = [
            "http::http_get",
            "http::http_post",
            "http::http_put",
            "http::http_patch",
            "http::http_delete",
            'format!("{base}/',
        ]
        for token in forbidden:
            if token in content:
                fail(
                    f"{path.relative_to(ROOT)} bypasses the portable operation boundary via {token}"
                )


def main() -> int:
    rust_source = (ROOT / "crates/ugoite-api-client/src/lib.rs").read_text()
    ts_source = (ROOT / "frontend/src/lib/ugoite-client/protocol.ts").read_text()
    rust_operations = rust_manifest_operations(rust_source)
    ts_operations = typescript_manifest_operations(ts_source)

    if len(rust_operations) != len(set(rust_operations)):
        fail("Rust operation manifest contains duplicates")
    if len(ts_operations) != len(set(ts_operations)):
        fail("TypeScript operation manifest contains duplicates")
    if rust_operations != ts_operations:
        rust_only = sorted(set(rust_operations) - set(ts_operations))
        ts_only = sorted(set(ts_operations) - set(rust_operations))
        fail(
            "Rust/TypeScript operation manifests differ: "
            f"rust_only={rust_only}, ts_only={ts_only}"
        )

    supported = set(rust_operations)
    for label, uses in [
        ("frontend", frontend_operation_uses()),
        ("CLI", cli_operation_uses()),
    ]:
        unknown = sorted(set(uses) - supported)
        if unknown:
            fail(f"{label} uses unknown portable operations: {unknown}")

    prepare_section = rust_source.split("pub fn prepare_request", 1)[1].split(
        "pub fn decode_response", 1
    )[0]
    decode_section = rust_source.split("fn operation_spec", 1)[1].split(
        "fn required_string", 1
    )[0]
    for operation in rust_operations:
        token = f'"{operation}" =>'
        if token not in prepare_section:
            fail(f"prepare_request is missing {operation}")
        if token not in decode_section:
            fail(f"operation_spec is missing {operation}")

    check_cargo_manifests()
    check_source_boundaries()

    print(
        "portable API client structural validation passed: "
        f"{len(rust_operations)} operations, "
        f"{len(frontend_operation_uses())} used by frontend, "
        f"{len(cli_operation_uses())} used by CLI"
    )
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except AssertionError as error:
        print(f"portable API client structural validation failed: {error}", file=sys.stderr)
        raise SystemExit(1)

File: scripts/check-portable-api-client.ts
#!/usr/bin/env -S deno run --allow-read

const ROOT = new URL("../", import.meta.url);

const fail = (message: string): never => {
  throw new Error(message);
};

const readText = async (relativePath: string): Promise<string> =>
  await Deno.readTextFile(new URL(relativePath, ROOT));

const quotedItems = (block: string): string[] =>
  [...block.matchAll(/"([a-z][a-z0-9_.]+)"/g)].map((match) => match[1]);

const rustManifestOperations = (source: string): string[] => {
  const match = source.match(
    /pub const SUPPORTED_OPERATIONS: &\[&str\] = &\[(.*?)\];/s,
  );
  if (!match) fail("Rust SUPPORTED_OPERATIONS manifest was not found");
  return quotedItems(match[1]);
};

const typescriptManifestOperations = (source: string): string[] => {
  const match = source.match(
    /export const UGOITE_API_OPERATIONS = \[(.*?)\] as const;/s,
  );
  if (!match) fail("TypeScript UGOITE_API_OPERATIONS manifest was not found");
  return quotedItems(match[1]);
};

const frontendOperationUses = async (): Promise<Map<string, string[]>> => {
  const pattern = /protocolFetch(?:<.*?>)?\s*\(\s*["']([^"']+)/gs;
  const uses = new Map<string, string[]>();
  for await (const entry of Deno.readDir(new URL("frontend/src/lib/", ROOT))) {
    if (!entry.isFile || !entry.name.endsWith("-api.ts")) continue;
    const relativePath = `frontend/src/lib/${entry.name}`;
    const content = await readText(relativePath);
    for (const match of content.matchAll(pattern)) {
      const operation = match[1];
      uses.set(operation, [...(uses.get(operation) ?? []), relativePath]);
    }
  }
  return uses;
};

const cliOperationUses = async (): Promise<Map<string, string[]>> => {
  const pattern = /http::execute\s*\(\s*[^,]+,\s*["']([^"']+)/gs;
  const uses = new Map<string, string[]>();
  for await (
    const entry of Deno.readDir(new URL("crates/ugoite-cli/src/commands/", ROOT))
  ) {
    if (!entry.isFile || !entry.name.endsWith(".rs")) continue;
    const relativePath = `crates/ugoite-cli/src/commands/${entry.name}`;
    const content = await readText(relativePath);
    for (const match of content.matchAll(pattern)) {
      const operation = match[1];
      uses.set(operation, [...(uses.get(operation) ?? []), relativePath]);
    }
  }
  return uses;
};

const dependencyKeysFromManifest = async (
  relativePath: string,
): Promise<Set<string>> => {
  const dependencies = new Set<string>();
  let inDependencies = false;
  const content = await readText(relativePath);
  for (const rawLine of content.split("\n")) {
    const line = rawLine.trim();
    if (line.startsWith("[")) {
      inDependencies = line === "[dependencies]";
      continue;
    }
    if (!inDependencies || !line || line.startsWith("#") || !line.includes("=")) {
      continue;
    }
    const [key] = line.split("=", 1);
    if (key.trim()) dependencies.add(key.trim());
  }
  return dependencies;
};

const checkCargoManifests = async (): Promise<void> => {
  for await (const entry of walk(new URL("./", ROOT))) {
    if (!entry.isFile || !entry.path.endsWith("Cargo.toml")) continue;
    await Deno.readTextFile(entry.path);
  }
  await readText("Cargo.lock");

  const apiDependencies = await dependencyKeysFromManifest(
    "crates/ugoite-api-client/Cargo.toml",
  );
  const forbiddenApi = new Set([
    "axum",
    "opendal",
    "reqwest",
    "tokio",
    "ugoite-core",
    "ugoite-domain",
    "ugoite-storage",
    "wasm-bindgen",
    "web-sys",
  ]);
  const unexpectedApi = [...apiDependencies].filter((dep) => forbiddenApi.has(dep))
    .sort();
  if (unexpectedApi.length > 0) {
    fail(`ugoite-api-client has forbidden dependencies: ${unexpectedApi.join(", ")}`);
  }

  const wasmDependencies = await dependencyKeysFromManifest(
    "crates/ugoite-wasm/Cargo.toml",
  );
  const forbiddenWasm = new Set([
    "opendal",
    "reqwest",
    "tokio",
    "ugoite-core",
    "ugoite-storage",
  ]);
  const unexpectedWasm = [...wasmDependencies].filter((dep) =>
    forbiddenWasm.has(dep)
  ).sort();
  if (unexpectedWasm.length > 0) {
    fail(`ugoite-wasm has forbidden dependencies: ${unexpectedWasm.join(", ")}`);
  }
};

const checkSourceBoundaries = async (): Promise<void> => {
  for await (const entry of Deno.readDir(new URL("frontend/src/lib/", ROOT))) {
    if (!entry.isFile || !entry.name.endsWith("-api.ts")) continue;
    const relativePath = `frontend/src/lib/${entry.name}`;
    const content = await readText(relativePath);
    if (content.includes("apiFetch")) {
      fail(`${relativePath} imports or calls apiFetch directly`);
    }
  }

  const forbiddenTokens = [
    "http::http_get",
    "http::http_post",
    "http::http_put",
    "http::http_patch",
    "http::http_delete",
    'format!("{base}/',
  ];
  for await (
    const entry of Deno.readDir(new URL("crates/ugoite-cli/src/commands/", ROOT))
  ) {
    if (!entry.isFile || !entry.name.endsWith(".rs")) continue;
    const relativePath = `crates/ugoite-cli/src/commands/${entry.name}`;
    const content = await readText(relativePath);
    for (const token of forbiddenTokens) {
      if (content.includes(token)) {
        fail(`${relativePath} bypasses the portable operation boundary via ${token}`);
      }
    }
  }
};

const walk = async function* (dir: URL): AsyncGenerator<Deno.DirEntry & {
  path: string;
}> {
  for await (const entry of Deno.readDir(dir)) {
    const path = new URL(entry.name, dir);
    if (entry.isDirectory) {
      yield* walk(new URL(`${entry.name}/`, dir));
      continue;
    }
    yield { ...entry, path: path.pathname };
  }
};

const ensureUnique = (label: string, values: string[]) => {
  if (values.length !== new Set(values).size) {
    fail(`${label} operation manifest contains duplicates`);
  }
};

const sectionBetween = (
  source: string,
  start: string,
  end: string,
  label: string,
): string => {
  const startIndex = source.indexOf(start);
  if (startIndex === -1) fail(`${label} start marker was not found`);
  const rest = source.slice(startIndex + start.length);
  const endIndex = rest.indexOf(end);
  if (endIndex === -1) fail(`${label} end marker was not found`);
  return rest.slice(0, endIndex);
};

const main = async (): Promise<void> => {
  const rustSource = await readText("crates/ugoite-api-client/src/lib.rs");
  const tsSource = await readText("frontend/src/lib/ugoite-client/protocol.ts");
  const rustOperations = rustManifestOperations(rustSource);
  const tsOperations = typescriptManifestOperations(tsSource);

  ensureUnique("Rust", rustOperations);
  ensureUnique("TypeScript", tsOperations);

  if (
    rustOperations.length !== tsOperations.length ||
    rustOperations.some((operation, index) => operation !== tsOperations[index])
  ) {
    const rustSet = new Set(rustOperations);
    const tsSet = new Set(tsOperations);
    const rustOnly = [...rustSet].filter((item) => !tsSet.has(item)).sort();
    const tsOnly = [...tsSet].filter((item) => !rustSet.has(item)).sort();
    fail(
      `Rust/TypeScript operation manifests differ: rust_only=${JSON.stringify(rustOnly)}, ts_only=${JSON.stringify(tsOnly)}`,
    );
  }

  const frontendUses = await frontendOperationUses();
  const cliUses = await cliOperationUses();
  const supported = new Set(rustOperations);
  for (const [label, uses] of [["frontend", frontendUses], ["CLI", cliUses]] as const) {
    const unknown = [...uses.keys()].filter((operation) => !supported.has(operation))
      .sort();
    if (unknown.length > 0) {
      fail(`${label} uses unknown portable operations: ${unknown.join(", ")}`);
    }
  }

  const prepareSection = sectionBetween(
    rustSource,
    "pub fn prepare_request",
    "pub fn decode_response",
    "prepare_request section",
  );
  const decodeSection = sectionBetween(
    rustSource,
    "fn operation_spec",
    "fn required_string",
    "operation_spec section",
  );
  for (const operation of rustOperations) {
    const token = `"${operation}" =>`;
    if (!prepareSection.includes(token)) fail(`prepare_request is missing ${operation}`);
    if (!decodeSection.includes(token)) fail(`operation_spec is missing ${operation}`);
  }

  await checkCargoManifests();
  await checkSourceBoundaries();

  console.log(
    `portable API client structural validation passed: ${rustOperations.length} operations, ${frontendUses.size} used by frontend, ${cliUses.size} used by CLI`,
  );
};

if (import.meta.main) {
  try {
    await main();
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`portable API client structural validation failed: ${message}`);
    Deno.exit(1);
  }
}
